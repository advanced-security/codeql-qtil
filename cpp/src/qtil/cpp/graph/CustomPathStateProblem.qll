private import qtil.locations.CustomPathStateProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path state problems.
import PathStateProblem<Location, CppLocatableConfig>
