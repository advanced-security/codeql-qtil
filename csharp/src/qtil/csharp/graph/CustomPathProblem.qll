private import qtil.locations.CustomPathProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path problems.
import PathProblem<Location, CsharpLocatableConfig>
