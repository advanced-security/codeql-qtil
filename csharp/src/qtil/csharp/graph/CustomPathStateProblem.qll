private import qtil.locations.CustomPathStateProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path state problems.
import PathStateProblem<Location, CsharpLocatableConfig>
