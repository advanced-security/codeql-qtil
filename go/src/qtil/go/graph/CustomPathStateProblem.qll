private import qtil.locations.CustomPathStateProblem
private import qtil.go.locations.Locatable
private import go
// Import the Go specific configuration for making custom path state problems.
import PathStateProblem<go::DbLocation, GoLocatableConfig>
