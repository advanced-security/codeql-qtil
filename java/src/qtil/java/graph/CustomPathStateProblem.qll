private import qtil.locations.CustomPathStateProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path state problems.
import PathStateProblem<Location, JavaLocatableConfig>
