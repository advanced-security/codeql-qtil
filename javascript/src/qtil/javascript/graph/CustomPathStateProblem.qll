private import qtil.locations.CustomPathStateProblem
private import qtil.javascript.locations.Locatable
private import javascript
// Import the javascript specific configuration for making custom path state problems.
import PathStateProblem<DbLocation, JavascriptLocatableConfig>