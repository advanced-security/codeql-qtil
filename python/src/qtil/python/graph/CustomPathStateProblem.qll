private import qtil.locations.CustomPathStateProblem
private import qtil.python.locations.Locatable
private import python
// Import the python specific configuration for making custom path state problems.
import PathStateProblem<Location, PythonLocatableConfig>
