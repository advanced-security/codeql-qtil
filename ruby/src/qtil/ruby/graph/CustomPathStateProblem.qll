private import qtil.locations.CustomPathStateProblem
private import qtil.ruby.locations.Locatable
private import ruby
// Import the Ruby specific configuration for making custom path state problems.
import PathStateProblem<Location, RubyLocatableConfig>