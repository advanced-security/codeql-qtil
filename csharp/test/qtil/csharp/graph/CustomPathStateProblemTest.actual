WARNING: unused variable 'depth' (CustomPathStateProblemTest.ql:13,29-34)
edges
| test.cs:3:13:3:15 | mid | test.cs:5:13:5:15 | end |  |  |
| test.cs:4:13:4:17 | start | test.cs:3:13:3:15 | mid |  |  |
nodes
| test.cs:3:13:3:15 | mid | semmle.label | mid |
| test.cs:4:13:4:17 | start | semmle.label | start |
| test.cs:5:13:5:15 | end | semmle.label | end |
#select
| test.cs:4:13:4:17 | start | test.cs:5:13:5:15 | end | Path from $@ (depth 0) to $@ (depth 2). | start | test.cs:4:13:4:17 | start | end | test.cs:5:13:5:15 | end |
