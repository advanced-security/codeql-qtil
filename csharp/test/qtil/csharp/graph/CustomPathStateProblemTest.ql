// Minimal test to make sure CustomPathStateProblem works with C# code.
import csharp
import qtil.csharp.graph.CustomPathStateProblem

module CallGraphPathStateProblemConfig implements CustomPathStateProblemConfigSig {
  class Node = LocalVariable;

  class State = int; // Track search depth

  predicate start(Node n, int depth) { n.getName() = "start" and depth = 0 }

  bindingset[depth]
  predicate end(Node n, int depth) { n.getName() = "end" and depth >= 0 }

  bindingset[depth1]
  bindingset[depth2]
  predicate edge(LocalVariable a, int depth1, LocalVariable b, int depth2) {
    depth2 = depth1 + 1 and
    exists(LocalVariableDeclExpr declA, LocalVariableDeclExpr declB |
      declA.getVariable() = a and
      declB.getVariable() = b and
      declA.getInitializer().(VariableAccess).getTarget() = b
    )
  }
}

import CustomPathStateProblem<CallGraphPathStateProblemConfig>

from LocalVariable start, LocalVariable end
where problem(start, end)
select start, end, "Path from $@ to $@.", start.getName(), start, end.getName(), end
