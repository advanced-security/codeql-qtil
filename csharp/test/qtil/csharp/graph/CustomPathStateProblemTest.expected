edges
| test.cs:4:13:4:15 | mid | test.cs:3:13:3:15 | end |  |  |
| test.cs:5:13:5:17 | start | test.cs:4:13:4:15 | mid |  |  |
nodes
| test.cs:3:13:3:15 | end | semmle.label | end |
| test.cs:4:13:4:15 | mid | semmle.label | mid |
| test.cs:5:13:5:17 | start | semmle.label | start |
#select
| test.cs:5:13:5:17 | start | test.cs:3:13:3:15 | end | Path from $@ to $@. | start | test.cs:5:13:5:17 | start | end | test.cs:3:13:3:15 | end |